ethereum/test/shutdownswitch.js

```diff
@@ -0,0 +1,471 @@
+            from: accounts[0],
+            gasLimit: 2000000
+        });
+    })
+
+    it("should reject redeems when they are disabled", async function () {
+        // This test assumes the previous test initialized the vaa variable.
+        assert.ok(vaa != null)
+
+        const accounts = await web3.eth.getAccounts();
+        const initialized = new web3.eth.Contract(BridgeImplementationFullABI, TokenBridge.address);
+        await clearAllVotes(initialized, accounts);
+        assert.equal((await initialized.methods.enabledFlag().call()), true)
+
+        // Redeem of this transfer should not be blocked by shutdown switch. It will instead fail with
+        // "no quorum", but that's okay for this test, since that check happens after the enable check.
+        try {
+            await initialized.methods.completeTransfer("0x" + vaa).send({
+                value: 0,
+                from: accounts[0],
+                gasLimit: 2000000
+            })
+        } catch (error) {
+            assert.equal(error.message, "Returned error: VM Exception while processing transaction: revert no quorum")
+        }
+
+        // Cast three votes to disable transfers.
+        await initialized.methods.castShutdownVote(false).send({
+            value: 0,
+            from: accounts[0],
+            gasLimit: 2000000
+        });
+
+        assert.equal((await initialized.methods.numVotesToDisable().call()), 1)
+        assert.equal((await initialized.methods.enabledFlag().call()), true)
+
+        await initialized.methods.castShutdownVote(false).send({
+            value: 0,
+            from: accounts[1],
+            gasLimit: 2000000
+        });
+
+        assert.equal((await initialized.methods.numVotesToDisable().call()), 2)
+        assert.equal((await initialized.methods.enabledFlag().call()), true)
+
+        await initialized.methods.castShutdownVote(false).send({
+            value: 0,
+            from: accounts[2],
+            gasLimit: 2000000
+        });
+
+        // Make sure transfers are now disabled.
+        assert.equal((await initialized.methods.numVotesToDisable().call()), 3)
+        assert.equal((await initialized.methods.enabledFlag().call()), false)
+
+        // This redeem should be blocked by shutdown switch rather than "no quorum".
+        try {
+            await initialized.methods.completeTransfer("0x" + vaa).send({
+                value: 0,
+                from: accounts[0],
+                gasLimit: 2000000
+            })
+        } catch (error) {
+            assert.equal(error.message, "Returned error: VM Exception while processing transaction: revert transfers are temporarily disabled")
+        }
+    })
+
+    it("a vote changing back to enabled should allow redeems again", async function () {
+        const accounts = await web3.eth.getAccounts();
+        const initialized = new web3.eth.Contract(BridgeImplementationFullABI, TokenBridge.address);
+        // Don't clear the votes here.
```

Comment:
in the interest in having tests not depend on each other's state, I'd almost suggest a call to `clearAllVotes()` and re-add (but without all the interim asserts from above).  This is mainly to try and avoid an unexpected state change in one of the other tests having a negative impact on this test.

Reply:
I had something like this bite me in an environment where you weren't exactly sure the order in which tests would execute, so you could get emergent pollution across tests.  My sense is that we're serial in this context atm.

Reply:
Yeah, the tests have to be serial, or at least not execute in parallel, or they will not work. Also, I put the checks at the start of this test to verify that the previous test worked before carrying on with this test. So unless you feel strongly about it, I would like to leave it as is.

Reply:
seems fine, more just seeing this for the first time, so please forgive my noob questions.